Dynamically generate carrier registration forms

harriswong opened the pull request:
## Description
This PR address bullet 2 on https://github.com/Automattic/woocommerce-services/issues/2234
> Building a registration form that is dynamically built based on the list of fields for each carrier (the UPS account registration form at /wp-admin/admin.php?page=wc-settings&tab=shipping&section=woocommerce-services-settings&carrier=UPS should be used as a design reference)

### Notes
- No form validation on this page. Every carrier will be different and we are not supplied with their individual validation rules.
- Carrier description is in `Lorem ipsum` placeholder text.

### Related issue(s)
Closes https://github.com/Automattic/woocommerce-services/issues/2234

### Steps to reproduce & screenshots/GIFs
1. Pull this branch, go to https://localhost/wp-admin/admin.php?page=wc-settings&tab=shipping&section=woocommerce-services-settings&carrier=DhlExpressAccount
2. You should see something like this:
https://localhost/wp-admin/admin.php?page=wc-settings&tab=shipping&section=woocommerce-services-settings&carrier=DhlExpressAccount
![image](https://user-images.githubusercontent.com/572862/98575441-7509e480-2276-11eb-9562-db24d7cefcfd.png)
3. Enter values into the form and press "Connect"
4. In the network tab, you should see the request sent with the form values you entered:
![image](https://user-images.githubusercontent.com/572862/98575737-c9ad5f80-2276-11eb-85c3-fb8abe2eba70.png)
Error is expected because the end point does not support this registration yet.
5. Click "Cancel" and it should bring you back to the previous page.

### Checklist

<!-- All testable code should have tests. -->
- [x] unit tests

<!-- An entry in the changelog file is always required -->
- [x] `changelog.txt` entry added

frosso commented on client/extensions/woocommerce/woocommerce-services/views/carrier-accounts/dynamic-settings.js:
```diff
@@ -0,0 +1,75 @@
+ */
+import React, { useEffect, useState } from 'react';
+import PropTypes from 'prop-types';
+import { connect } from 'react-redux';
```
Too bad we're using `react-redux@5.*`. If we were using `7.1`, we would have been able to use hooks also on this :)

frosso commented on client/extensions/woocommerce/woocommerce-services/views/carrier-accounts/dynamic-settings-form.js:
```diff
@@ -0,0 +1,199 @@
+        setFormFields(formFieldsCopy);
+    };
+
+    const getInputComponent = (fieldKey) => {
```
What do you think about moving the `getInputComponent` to a component factory? This allows React to better manage the component tree and its updates, without having to re-generate the components each time.

frosso commented on client/extensions/woocommerce/woocommerce-services/views/carrier-accounts/dynamic-settings-form.js:
```diff
@@ -0,0 +1,199 @@
+        return formComponent;
+    };
+
+    const submitCarrierSettingsHandler = async () => {
```
Having an `async` handler can have some performance implications. This blocks the browser and "freezes" the UI until the promise is resolved.

I recommend using a little trick to prevent this: declare the async function _inside_ the handler and call it _from_ the handler.

An example is added to the other comment with all the code suggestions.

harriswong replied:
Good callout! This reminds me of a eslint warning thrown by https://github.com/facebook/react/blob/369c3db6299506d9d1ab6628accc2c40827d1f1f/packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js#L103-L113. I didn't know it would freezes the UI. 

Maybe we should add eslint-plugin-react-hooks to our eslint rules. There are some useful things for hooks.

frosso commented on client/extensions/woocommerce/woocommerce-services/views/carrier-accounts/dynamic-settings-form.js:
```diff
@@ -0,0 +1,199 @@
+    const [formFields, setFormFields] = useState({});
+    const [isSaving, setIsSaving] = useState(false);
+
+    const updateValue = ( fieldKey ) => ( newValue ) => {
```
I'd recommend memoizing most (if not all) of these handlers. Not memoizing them has some implications on React's ability to perform DOM updates. When memoized, React can compare the previous and next values and decide whether or not the component tree needs to update.
When it's not memoized, React just decides to update the whole tree (because the handler's ref changes at each re-render).

frosso commented on client/extensions/woocommerce/woocommerce-services/views/carrier-accounts/dynamic-settings-form.js:
```diff
@@ -0,0 +1,199 @@
+    translate: PropTypes.func,
+};
+
+export default connect( mapStateToProps, mapDispatchToProps )( localize( DynamicCarrierAccountSettingsForm ) );
```
All the suggestions above (unless I missed some) are in this big ol' chunk:

```
/**
 * External dependencies
 */
import React, { useState, useCallback } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { localize } from 'i18n-calypso';

/**
 * Internal dependencies
 */
import Button from 'components/button';
import CompactCard from 'components/card/compact';
import Checkbox from 'woocommerce/woocommerce-services/components/checkbox';
import TextField from 'woocommerce/woocommerce-services/components/text-field';
import { getSelectedSiteId } from 'state/ui/selectors';
import * as api from 'woocommerce/woocommerce-services/api';
import { errorNotice, successNotice } from 'state/notices/actions';

const FormFieldFactory = ( { visibility, label, id, value, onChange } ) => {
	const handleToggle = useCallback(() => {
		onChange(id, !value);
	}, [id, value, onChange]);

	const handleTextChange = useCallback(( newValue ) => {
		onChange(id, newValue);
	}, [id, value, onChange]);

	switch (visibility) {
		case 'select':
		case 'checkbox':
			return (
				<>
					<Checkbox
						id={ id }
						onChange={ handleToggle }
						checked={ value || false }
					/>
					<label htmlFor={ id }><span>{ label }</span></label>
				</>
			);

		case 'invisible':
		case 'masked':
		case 'readonly':
		// TODO: We will need to handle the above invisible/readonly fields separately. For now, use default.
		case 'password':
		// TODO: We will need to create a PasswordField component, for now, use default.
		case 'visible':
		default:
			return (
				<TextField
					id={ id }
					title={ label }
					updateValue={ handleTextChange }
					value={ value || '' }
				/>
			);
	}
}

export const DynamicCarrierAccountSettingsForm = ( props ) => {
	const {
        translate,
		siteId,
		carrierType,
		noticeActions,
		carrierName,
    } = props;

    const [formValues, setFormValues] = useState({});
    const [isSaving, setIsSaving] = useState(false);

    const handleFormFieldChange = useCallback((id, newValue) => {
    	setFormValues((oldValues) => ({
    		...oldValues,
		    [id]: newValue,
	    }))
    }, [setFormValues]);

	const handleSubmit = useCallback( () => {
		const submit = async () => {
			setIsSaving(true);

			try {
				await api.post( siteId, api.url.shippingCarrier(), {
					type: carrierType,
					settings: formValues
				} );
			} catch (error) {
				noticeActions.errorNotice(`Failed to register the carrier. ${ error }`);
			}

			setIsSaving(false);
		}

		submit();
	}, [isSaving, setIsSaving, siteId, carrierType, formValues, noticeActions] );

    const handleCancel = useCallback(() => {
        history.back();
    }, []);

    return (
        <div className="carrier-accounts__settings-container">
			<div className="carrier-accounts__settings">
				<div className="carrier-accounts__settings-info">
					<h4 className="carrier-accounts__settings-subheader-above-description">
						{ translate( 'Connect your %(carrierName)s account', {
                            args: {
                                carrierName,
                            }
                        } ) }
					</h4>
					<p className="carrier-accounts__settings-subheader-description">
                    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam et dolor quam.
					</p>
				</div>
				<div className="carrier-accounts__settings-form">
                    <CompactCard>
                        <h4 className="carrier-accounts__settings-subheader">General Information</h4>
                        <p className="carrier-accounts__settings-subheader-description">
                            { translate( 'This is the account number and address from your %(carrierName)s profile', {
                            args: {
                                carrierName,
                            }
                        } ) }
                        </p>
                    </CompactCard>

                    <CompactCard>
                        {props.registrationFields && Object.entries(props.registrationFields).map( ( [key, field] ) => (
                        	<FormFieldFactory key={key} id={key} visibility={field.visibility} label={field.label} value={formValues[key]} onChange={handleFormFieldChange} />
                        ))}
                    </CompactCard>

                    <CompactCard className="carrier-accounts__settings-actions">
						<Button
							compact
							primary
                            onClick={ handleSubmit }
                            disabled={ isSaving }
                        >
							{ translate( 'Connect' ) }
						</Button>
                        <Button
                            compact
                            onClick={ handleCancel }
                        >
							{ translate( 'Cancel' ) }
						</Button>
					</CompactCard>
                </div>
            </div>
        </div>
	);
};

const mapStateToProps = ( state ) => {
	return {
		siteId: getSelectedSiteId( state ),
    };
};

const mapDispatchToProps = ( dispatch ) => ({
	noticeActions: bindActionCreators( { successNotice, errorNotice }, dispatch ),
});

DynamicCarrierAccountSettingsForm.propTypes = {
    carrierType: PropTypes.string,
    carrierName: PropTypes.string,
    registrationFields: PropTypes.object,
    siteId: PropTypes.number,
    translate: PropTypes.func,
};

export default connect( mapStateToProps, mapDispatchToProps )( localize( DynamicCarrierAccountSettingsForm ) );
```

harriswong replied:
Codes worth 1000 words! Thanks for the explanation. The `<FormFieldFactory/>` makes a lot of sense so that React takes advantage of its own component managing, versus a loop that dumps out HTMLs.

frosso replied:
Thanks for reading through this :)

harriswong replied:
I have decided to use this whole snippets. ;)

frosso commented on client/extensions/woocommerce/woocommerce-services/views/carrier-accounts/dynamic-settings.js:
```diff
@@ -0,0 +1,75 @@
+
+	const supportedCarriers = carrierRegistrationFields.filter(carrier => carrier.type === props.carrier);
+
+	if (!supportedCarriers || supportedCarriers.length === 0) {
```
In which case `supportedCarriers` would be falsy? `Array.filter` always returns an array.
```suggestion
	if (!supportedCarriers.length === 0) {
```

frosso commented on client/extensions/woocommerce/woocommerce-services/views/carrier-accounts/dynamic-settings.js:
```diff
@@ -0,0 +1,75 @@
+		);
+	}
+
+	const currentCarrierRegistrationField = supportedCarriers[0];
```
(minor) I think that L39 could just become 

```
const [ currentCarrierRegistrationField ] = carrierRegistrationFields.filter(carrier => carrier.type === props.carrier);
```

For some syntactic sugar :)

harriswong replied:
I quite like sweets.

frosso commented on client/extensions/woocommerce/woocommerce-services/views/carrier-accounts/dynamic-settings.js:
```diff
@@ -0,0 +1,75 @@
+};
+
+const mapDispatchToProps = ( dispatch, {siteId} ) => ( {
+	getCarrierRegistrationFields: () => dispatch( getCarrierRegistrationFields( siteId ) ),
```
There should be no need to connect this, since it's not dispatching any Redux actions.

harriswong replied:
Yes, we should not need this anymore. Thanks.

frosso commented on client/extensions/woocommerce/woocommerce-services/views/carrier-accounts/test/dynamic-settings-form.js:
```diff
@@ -0,0 +1,119 @@
+jest.mock('components/forms/form-text-input', () => {
+    return function DummyFormTextField(props) {
+        return (
+        <div>
```
(minor) looks like spacing is borked in a few places

harriswong replied:
It looks like our eslint rule doesn't check for spacing. We should prob fix this.

harriswong replied:
Fixed. Converted space to tabs.

frosso reviewed (commented):
The functionality looks good! I'm leaving a few comments with some suggestions on the hooks' functionality

harriswong commented:
@frosso Great feedback, thank you!! I am going to address the comments and put this back to "ready for review" once done.

harriswong commented on client/extensions/woocommerce/woocommerce-services/state/carrier-accounts/actions.js:
```diff
@@ -93,3 +93,8 @@ export const submitCarrierSettings = ( siteId, carrier, values ) => ( dispatch )
 };
+
+export const getCarrierRegistrationFields = ( siteId ) => () => {
+	return api.get(siteId, api.url.shippingCarrierTypes())
```
I am pretty sure this is no longer needed. This is remnants prior of me switching from Redux store to component stats via hooks. The `api.url.shippingCarrierTypes()` data is now directly read from the `useEffect` hooks. I am going to remove this.

harriswong commented:
@frosso Please take a look again.

- I rebased latest `develop`
- I addressed the comments and applied your suggestions.
- Fixed test
- Lint fixes

frosso commented on client/extensions/woocommerce/woocommerce-services/views/carrier-accounts/dynamic-settings.js:
```diff
@@ -0,0 +1,66 @@
+	siteId: PropTypes.number
+};
+
+export default connect( mapStateToProps )( localize( DynamicCarrierAccountSettings ) );
```
The props injected by the `localize` HOC are not used by this component
```suggestion
export default connect( mapStateToProps )( DynamicCarrierAccountSettings );
```

harriswong replied:
Good call. We should localizing. I am going to wrap text with `translate`.

frosso reviewed (approved):
There's still a few instances I didn't comment on where spacing is inconsistent, but overall looks good! Thanks for considering the feedback!

harriswong commented on client/apps/shipping-settings/view-wrapper.js:
```diff
@@ -69,7 +70,22 @@ class LabelSettingsWrapper extends Component {
 		const { carrier, carriers, isSaving, siteId, translate } = this.props;
 
-		if ( carrier ) {
+		if (!carrier) {
```
I am going to add an issue to cool-down to add spacing to our eslint (if there isn't one already)

harriswong commented:
@frosso Please take a look again
- I fixed some linting issues you found
- I added translate() to the text
- Updated tests

frosso reviewed (approved):
Thanks for the changes! 💯
